consensus/hotstuff/timeoutcollector/timeout_processor_test.go

```diff
@@ -0,0 +1,502 @@
+	s.notifier.On("OnPartialTcCreated", s.view).Return(nil).Once()
+	s.notifier.On("OnTcConstructedFromTimeouts", mock.Anything).Run(func(args mock.Arguments) {
+		newestQC := timeouts[len(timeouts)-1].NewestQC
+		tc := args.Get(0).(*flow.TimeoutCertificate)
+		// ensure that TC contains correct fields
+		expectedTC := &flow.TimeoutCertificate{
+			View:          s.view,
+			NewestQCViews: highQCViews,
+			NewestQC:      newestQC,
+			SignerIndices: signerIndices,
+			SigData:       expectedSig,
+		}
+		require.Equal(s.T(), expectedTC, tc)
+	}).Return(nil).Once()
+
+	s.sigAggregator.On("Aggregate").Return([]flow.Identifier(signers.NodeIDs()), highQCViews, expectedSig, nil)
+	s.committee.On("IdentitiesByEpoch", s.view).Return(s.participants, nil)
+
+	for _, timeout := range timeouts {
+		err := s.processor.Process(timeout)
+		require.NoError(s.T(), err)
+	}
+	s.notifier.AssertExpectations(s.T())
+	s.sigAggregator.AssertExpectations(s.T())
+
+	// add extra timeout, make sure we don't create another TC
+	// should be no-op
+	timeout := helper.TimeoutObjectFixture(
+		helper.WithTimeoutObjectView(s.view),
+		helper.WithTimeoutNewestQC(helper.MakeQC(helper.WithQCView(lastSuccessfulQC.View))),
+		helper.WithTimeoutObjectSignerID(s.participants[0].NodeID),
+		helper.WithTimeoutLastViewTC(nil),
+	)
+	err = s.processor.Process(timeout)
+	require.NoError(s.T(), err)
+
+	s.notifier.AssertExpectations(s.T())
+	s.validator.AssertExpectations(s.T())
+}
+
+// TestProcess_ConcurrentCreatingTC tests a scenario where multiple goroutines process timeout at same time,
+// we expect only one TC created in this scenario.
+func (s *TimeoutProcessorTestSuite) TestProcess_ConcurrentCreatingTC() {
+	s.validator.On("ValidateQC", mock.Anything).Return(nil)
+	s.notifier.On("OnPartialTcCreated", mock.Anything).Return(nil).Once()
+	s.notifier.On("OnTcConstructedFromTimeouts", mock.Anything).Return(nil).Once()
+	s.sigAggregator.On("Aggregate").Return([]flow.Identifier(s.participants.NodeIDs()), []uint64{}, crypto.Signature{}, nil)
+	s.committee.On("IdentitiesByEpoch", mock.Anything).Return(s.participants, nil)
+	s.committee.On("IdentityByEpoch", mock.Anything, mock.Anything).Return(s.signer, nil)
+
+	var startupWg, shutdownWg sync.WaitGroup
+
+	newestQC := helper.MakeQC(helper.WithQCView(s.view - 1))
+
+	startupWg.Add(1)
+	// prepare goroutines, so they are ready to submit a timeout at roughly same time
+	for i, signer := range s.participants {
+		shutdownWg.Add(1)
+		timeout := helper.TimeoutObjectFixture(
+			helper.WithTimeoutObjectView(s.view),
+			helper.WithTimeoutNewestQC(newestQC),
+			helper.WithTimeoutObjectSignerID(signer.NodeID),
+			helper.WithTimeoutLastViewTC(nil),
+		)
+		go func(i int, timeout *model.TimeoutObject) {
+			defer shutdownWg.Done()
+			startupWg.Wait()
+			err := s.processor.Process(timeout)
+			require.NoError(s.T(), err)
+		}(i, timeout)
+	}
+
+	startupWg.Done()
+
+	// wait for all routines to finish
+	shutdownWg.Wait()
```

Comment:
```suggestion
	shutdownWg.Wait()
	
	s.notifier.AssertExpectations(s.T())
```
⚠️  I think we need this, otherwise we aren't sure whether the TC was actually created

Reply:
That's handy - Didn't know about that feature 👌